src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2148 @@
+            auto const preAlice = env.balance(alice);
+            auto const preAliceUSD = env.balance(alice, USD.issue());
+            auto const preBobSeq = env.seq(bob);
+            auto const preBob = env.balance(bob);
+            auto const preBobUSD = env.balance(bob, USD.issue());
+
+            auto const batchFee = calcBatchFee(env, 1, 2);
+            env(batch::batch(alice, preAliceSeq, batchFee, tfAllOrNothing),
+                batch::add(pay(alice, bob, XRP(10)), preAliceSeq + 1),
+                batch::add(pay(bob, alice, XRP(5)), preBobSeq + 10),
+                batch::sig(bob),
+                ter(tesSUCCESS));
+            TxID const batchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {};
+            env.close();
+            validateBatch(env, batchId, testCases);
+
+            // Alice pays fee & Bob should not be affected.
+            BEAST_EXPECT(env.seq(alice) == preAliceSeq + 1);
+            BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+            BEAST_EXPECT(env.balance(alice, USD.issue()) == preAliceUSD);
+            BEAST_EXPECT(env.seq(bob) == preBobSeq);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+            BEAST_EXPECT(env.balance(bob, USD.issue()) == preBobUSD);
+        }
+
+        // Invalid: Outer and Inner Sequence are the same
+        {
+            auto const preAliceSeq = env.seq(alice);
+            auto const preAlice = env.balance(alice);
+            auto const preAliceUSD = env.balance(alice, USD.issue());
+            auto const preBobSeq = env.seq(bob);
+            auto const preBob = env.balance(bob);
+            auto const preBobUSD = env.balance(bob, USD.issue());
+
+            auto const batchFee = calcBatchFee(env, 1, 2);
+            env(batch::batch(alice, preAliceSeq, batchFee, tfAllOrNothing),
+                batch::add(pay(alice, bob, XRP(10)), preAliceSeq),
+                batch::add(pay(bob, alice, XRP(5)), preBobSeq),
+                batch::sig(bob),
+                ter(tesSUCCESS));
+            TxID const batchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {};
+            env.close();
+            validateBatch(env, batchId, testCases);
+
+            // Alice pays fee & Bob should not be affected.
+            BEAST_EXPECT(env.seq(alice) == preAliceSeq + 1);
+            BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+            BEAST_EXPECT(env.balance(alice, USD.issue()) == preAliceUSD);
+            BEAST_EXPECT(env.seq(bob) == preBobSeq);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+            BEAST_EXPECT(env.balance(bob, USD.issue()) == preBobUSD);
+        }
+    }
+
+    void
+    testBadFeeOuterBatch(FeatureBitset features)
+    {
+        testcase("bad fee outer batch");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        // Bad Fee Without Signer
```

Comment:
Some additional tests I'd like to see on the fee:
* `tx.Signers` and `tx.BatchSigners`
* `txBatchSigners.Signers` exists